onnxruntime/core/providers/cuda/reduction/reduction_functions.cc

```diff
@@ -13,45 +15,86 @@
 namespace onnxruntime {
 namespace cuda {
 
-ApplicableMatrixReduction get_applicable_matrix_reduction(
-    const cudnnReduceTensorOp_t cudnn_reduce_op,
-    const std::vector<int64_t>& dims, const std::vector<int64_t>& original_axes,
-    int& m_out, int& n_out) {
-  if (cudnn_reduce_op != CUDNN_REDUCE_TENSOR_ADD) {
-    return ApplicableMatrixReduction::None;
+namespace {
+// gets min and max of single contiguous range of axes if available
+optional<std::pair<int64_t, int64_t>> GetMinMaxContiguousAxes(
+    int64_t rank,
+    const std::vector<int64_t>& dims,
+    const std::vector<int64_t>& original_axes) {
+  // empty axes means reduce all dimensions
+  if (original_axes.empty()) {
+    return std::make_pair(int64_t{0}, rank - 1);
   }
 
-  const auto rank = gsl::narrow<int64_t>(dims.size());
+  // normalize axis values and sort
+  const std::vector<int64_t> axes = [&original_axes, rank]() {
+    std::vector<int64_t> result(original_axes);
+    std::for_each(
+        result.begin(), result.end(),
+        [rank](int64_t& axis) { axis = HandleNegativeAxis(axis, rank); });
+    std::sort(result.begin(), result.end());
+    return result;
+  }();
+
+  assert(!axes.empty());
+
+  const auto is_dim_one = [](int64_t dim) { return dim == 1; };
 
-  // min and max of single contiguous range of axes
-  const auto minmax_axes = [&]() -> optional<std::pair<int64_t, int64_t>> {
-    // empty axes means reduce all dimensions
-    if (original_axes.empty()) {
-      return std::make_pair(int64_t{0}, rank - 1);
+  for (auto a = axes.begin(), b = axes.begin() + 1;
+       b != axes.end();
+       ++a, ++b) {
+    ORT_ENFORCE(*a != *b, "axes must not contain duplicate values");
+
+    // if axis values are adjacent, the axes are contiguous
+    if (*a + 1 == *b) {
+      continue;
     }
 
-    // normalize axis values and sort
-    const std::vector<int64_t> axes = [&original_axes, rank]() {
-      std::vector<int64_t> result(original_axes);
-      std::for_each(
-          result.begin(), result.end(),
-          [rank](int64_t& axis) { axis = HandleNegativeAxis(axis, rank); });
-      std::sort(result.begin(), result.end());
-      return result;
-    }();
-
-    for (auto a = axes.begin(), b = axes.begin() + 1;
-         b != axes.end();
-         ++a, ++b) {
-      ORT_ENFORCE(*a != *b, "axes must not contain duplicate values");
-      if (*a + 1 != *b) {  // not contiguous
-        return {};
-      }
+    // if all dimension values between adjacent axes are 1,
+    // treat the axes as contiguous
+    if (std::all_of(dims.begin() + *a + 1, dims.begin() + *b, is_dim_one)) {
+      continue;
     }
 
-    return std::make_pair(axes.front(), axes.back());
+    // otherwise, not contiguous
+    return nullopt;
```

Comment:
>nullopt [](start = 11, length = 7)

Not sure what's the convention here. Would it be `std::nullopt` or just `nullopt`.

Reply:
i'd prefer std::nullopt, but we don't have C++17's std::optional so this is actually onnxruntime::nullopt

---
In reply to: [520194944](https://github.com/microsoft/onnxruntime/pull/5734#discussion_r520194944) [](ancestors = 520194944)